pkg/cleanup/controller/controller.go

```diff
@@ -0,0 +1,251 @@
+
+	// delay is the amount of time to wait after Node deletion to cleanup resources.
+	delay time.Duration
+
+	// stalePVDiscoveryInterval is how often to scan for and delete PVs with affinity to a deleted Node.
+	stalePVDiscoveryInterval time.Duration
+}
+
+func NewCleanupController(client kubernetes.Interface, pvInformer coreinformers.PersistentVolumeInformer, nodeInformer coreinformers.NodeInformer, storageClassName string, delay time.Duration, stalePVDiscoveryInterval time.Duration) *CleanupController {
+	broadcaster := record.NewBroadcaster()
+	eventRecorder := broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: fmt.Sprintf("cleanup-controller")})
+
+	controller := &CleanupController{
+		client:           client,
+		storageClassName: storageClassName,
+		// Delayed queue with rate limiting
+		entryQueue: workqueue.NewRateLimitingQueueWithConfig(
+			workqueue.DefaultControllerRateLimiter(),
+			workqueue.RateLimitingQueueConfig{
+				Name: "stale",
+			}),
+		nodeLister:               nodeInformer.Lister(),
+		nodeListerSynced:         nodeInformer.Informer().HasSynced,
+		pvLister:                 pvInformer.Lister(),
+		pvListerSynced:           pvInformer.Informer().HasSynced,
+		eventRecorder:            eventRecorder,
+		broadcaster:              broadcaster,
+		delay:                    delay,
+		stalePVDiscoveryInterval: stalePVDiscoveryInterval,
+	}
+
+	// Set up event handler for when Nodes are deleted
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: controller.nodeDeleted,
+	})
+
+	return controller
+}
+
+// Run will start worker threads that try to process items off of the entryQueue, as well
+// as syncing informer caches and continuously running the Deleter.
+// It will block until the context gives a Done signal, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *CleanupController) Run(ctx context.Context, workers int) error {
+	defer utilruntime.HandleCrash()
+	defer c.entryQueue.ShutDown()
+
+	klog.Info("Starting to Run CleanupController")
+
+	c.broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: c.client.CoreV1().Events(v1.NamespaceAll)})
+	defer c.broadcaster.Shutdown()
+
+	klog.Info("Waiting for informer caches to sync")
+	// Wait for the caches to be synced before starting workers
+	if ok := cache.WaitForCacheSync(ctx.Done(), c.nodeListerSynced, c.pvListerSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.Infof("Starting workers, count: &d", workers)
+	// Launch workers to process
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	klog.Info("Started workers")
+
+	// Look for stale PVs and start timers for resource cleanup
+	c.startCleanupTimersIfNeeded()
+
+	// Run Deleter and block until channel is closed
+	deleter := deleter.NewDeleter(c.client, c.pvLister, c.nodeLister, c.storageClassName)
```

Comment:
I'd prefer making this piece of code into another Run() function in Deleter.go because decouplation give better scalability and testability.